### ISSUE TITLE
Last batch always pad up to the batch size and cause error "logits and labels must be broadcastable"

### ISSUE THREAD

**Issue by acmilannesta** [tensorflow/tensorflow#47894]
I'm using tensorflow 2.4.1 on Win 10 and Python 3.8.6.

When I use data generator and reach the last batch of training, the model will always pad up to the full batch size even if the real size is less than that. 

This will cause an error "logits and labels must be broadcastable", which is basically a mismatch of my label size (13x24) and predication size (32x24), which 32 is my batch size and 24 is number of classes. 

```
BATCH_SIZE = 32
NUM_EPOCHS = 1
NUM_CLASSES = 24
TIME_STEPS = 1000

def build_model():
    total = ((len(tr_idx) + BATCH_SIZE - 1) // BATCH_SIZE) * NUM_EPOCHS

    inp1 = layers.Input(shape=(TIME_STEPS, 12))
    # inp2 = layers.Input(shape=2)

    out = layers.BatchNormalization(axis=1)(inp1)
    out = layers.Conv1D(filters=64, kernel_size=50, strides=10, activation='relu')(out)
    out = layers.Conv1D(filters=64, kernel_size=50, strides=10, activation='relu')(out)
    out = layers.MaxPool1D(pool_size=2)(out)
    out = layers.Flatten()(out)
    out = layers.Dense(NUM_CLASSES*2, activation='relu')(out)

    out1 = layers.Dense(NUM_CLASSES, activation='softmax')(out)
    out2 = layers.Dense(NUM_CLASSES, activation='softmax')(out)
    out3 = layers.Dense(NUM_CLASSES, activation='softmax')(out)

    model = models.Model(inp1, [out1, out2, out3])
    model.compile(loss='categorical_crossentropy', optimizer=optimizers.Adam(learning_rate=keras.experimental.LinearCosineDecay(5e-4, total)))
    # model.summary()
    return model

model = build_model()
```

Here is my data generator
```
class datagen:
    def __init__(self, data, batch_size = BATCH_SIZE, branch='train'):
        self.data = data.reset_index(drop=True)
        self.batch_size = batch_size
        self.branch = branch
        self.steps = len(self.data) // self.batch_size
        if len(self.data) % self.batch_size !=0:
            self.steps += 1

    def __len__(self):
        return self.steps

    def __iter__(self):
        while True:
            if self.branch == 'train':
                self.data = self.data.sample(frac=1).reset_index(drop=True)
            for i in range(self.steps):
                batch_data = self.data[i * self.batch_size: (i+1) * self.batch_size]
                batch_image_ids = batch_data['ID'].tolist()
                with zipfile.ZipFile('record_step1k.zip', 'r') as zipf:
                    tmp = np.zeros((self.batch_size, TIME_STEPS, 12))
                    for cnt, id in enumerate(batch_image_ids):
                        x = zipf.read(id)
                        tmp[cnt] = np.frombuffer(x).reshape((12, -1)).transpose()
                if self.branch == 'train':
                    yield tmp, [keras.utils.to_categorical(batch_data['Dx_1'], NUM_CLASSES), 
                                keras.utils.to_categorical(batch_data['Dx_2'], NUM_CLASSES),
                                keras.utils.to_categorical(batch_data['Dx_3'], NUM_CLASSES)]
                else:
                    yield tmp
```

**Comment by acmilannesta** [comment]
Found the issue. In the data generator should change ``` tmp = np.zeros((self.batch_size, TIME_STEPS, 12))``` to ``` tmp = np.zeros((len(batch_data), TIME_STEPS, 12))```
